### ISSUE TITLE
Allow for composition based schemas

### ISSUE THREAD

**Issue by badrobit** [marshmallow-code/marshmallow#1339]
I have a series of schema's some with a variety of fields that are commonly shared but they build on each other in a manner like this: 

```python
class CommandType1(Schema): 
    field_1 = fields.Str() 
    field_2 = fields.Str()
    field_3 = fields.Str()

class CommandType2(Schema1): 
    field_4 = fields.Integer()

class CommandType3(Schema1): 
    field_5 = fields.Float()

class CommandType4(Schema2): 
    field_6 = fields.Str()
    field_7 = fields.Str()
    field_8 = fields.Str()

class CommandType5(Schema3): 
    field_6 = fields.Str()
    field_7 = fields.Str()
    field_8 = fields.Str()

# ... and on a few more times...
```
(note the fields here are just the quickest ones I could think of the real ones also contain validation checking to ensure the correctness of the field to the application in question)

These schemas represent commands that come into a system as JSON. As you can hopefully (now) see `CommandType4` and `CommandType5` are effectively the exact same type but have two shared parent commands with a different data type, they also act as a base command for other commands that would be implemented.

I was hoping there was a way of putting schema's together using composition (like in an Entity Component System fashion) instead of more classical inheritance. 

The goal would be to define a set of schemas that can be put together to avoid a large amount of duplicate code. A lot of the commands tend to come with sets of arguments that are required but also end up being required base components to more complex system commands. I know I could use nesting but a portion of the system cannot handled nested JSON so we need to keep everything within a single non-nested JSON string.

**Comment by sloria** [comment]
I'm not sure I understand the question. Schema4 and Schema5 don't share fields with their ancestors, so why would you need to combined the validators? Can you give an example of the expected input and output? Or maybe pseudocode?

**Comment by badrobit** [comment]
Hi, sorry for the rushed question I have updated the original to hopefully try and be a bit clearer,

**Comment by deckar01** [comment]
The mixin pattern with multiple inheritance should allow just about any type of schema composition.

https://stackoverflow.com/questions/533631/what-is-a-mixin-and-why-are-they-useful

```py
class CommandMixin: 
    field_6 = fields.Str()
    field_7 = fields.Str()
    field_8 = fields.Str()

class CommandType4(CommandMixin, Schema2): 
    pass

class CommandType5(CommandMixin, Schema3): 
    pass
```

**Comment by sloria** [comment]
Not sure if it fits your use case, but you could also dynamically generate a schema.

```python
from marshmallow import Schema, fields


class BaseCommandSchema(Schema):
    field_1 = fields.Str()
    field_2 = fields.Str()
    field_3 = fields.Str()


args = {"command2", "command4"}

additional_fields = {}
if "command2" in args:
    additional_fields.update({"field_4": fields.Str()})

    if "command4" in args:
        additional_fields.update(
            {"field_6": fields.Str(), "field_7": fields.Str(), "field_8": fields.Str()}
        )

CommandSchema = BaseCommandSchema.from_dict(additional_fields)
print(sorted(CommandSchema().fields.keys()))
# ['field_1', 'field_2', 'field_3', 'field_4', 'field_6', 'field_7', 'field_8']
```

Note: `from_dict` isn't yet released (it will be in 3.0 final), so you could add it to `BaseCommandSchema` for now.

Closing for now, feel free to carry on discussing here if your use case hasn't been met.
